modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshot.java

```diff
@@ -133,23 +140,61 @@ public PartitionKey partitionKey() {
     }
 
     /**
-     * Freezes the scope of this snapshot.
+     * Freezes the scope of this snapshot. This includes taking snapshot metadata and opening TX data cursor.
      *
      * <p>Must be called under snapshot lock.
      */
     void freezeScope() {
         assert mvOperationsLock.isLocked() : "MV operations lock must be acquired!";
 
+        frozenMeta = takeSnapshotMeta();
+
         txDataCursor = partition.txStatePartitionStorage().scan();
     }
 
+    private SnapshotMeta takeSnapshotMeta() {
+        long lastAppliedIndex = Math.max(
+                partition.mvPartitionStorage().lastAppliedIndex(),
+                partition.txStatePartitionStorage().lastAppliedIndex()
+        );
+
+        return SnapshotMetaUtils.snapshotMetaAt(lastAppliedIndex, logManager);
+    }
+
+    /**
+     * Returns metadata corresponding to this snapshot.
+     *
+     * @return This snapshot metadata.
+     */
+    public SnapshotMeta meta() {
+        SnapshotMeta meta = frozenMeta;
+
+        assert meta != null : "No snapshot meta yet, probably the snapshot scope was not yet frozen";
+
+        return meta;
+    }
+
     /**
-     * Reads a snapshot meta and returns a future with the response.
+     * Reads the snapshot meta and returns a future with the response.
      *
-     * @param metaRequest Meta request.
+     * @param request Meta request.
      */
-    SnapshotMetaResponse handleSnapshotMetaRequest(SnapshotMetaRequest metaRequest) {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17935
+    @Nullable
+    SnapshotMetaResponse handleSnapshotMetaRequest(SnapshotMetaRequest request) {
+        if (closed) {
+            return logAlreadyClosedAndReturnNull();
+        }
+
+        SnapshotMeta meta = frozenMeta;
+
+        assert meta != null : "No snapshot meta yet, probably the snapshot scope was not yet frozen";
+
+        return MESSAGES_FACTORY.snapshotMetaResponse().meta(meta).build();
+    }
+
+    @Nullable
+    private <T> T logAlreadyClosedAndReturnNull() {
```

Comment:
Missing the assertion that the method was called by `closed`.

Reply:
I don't understand what you suggest here. Please elaborate.

Reply:
What should the method do?

Reply:
It should log that someone is trying to get data from an already closed snapshot and then return `null`; callers handle that `null` gracefully.

Reply:
So let's check that this method was called after knowing that we closed the snapshot.

Reply:
This does not seem necessary. The method is always closed under `if (closed)`

Reply:
@tkalkirill , in this case I side with Roman, assertion here is excessive.
Two simple arguments from me:

1. This method is merely a utility method to save a line of code in several places, it doesn't do something that may put stability of the system at risk.
2. The method is private and context of its use is completely enclosed inside a rather short class. Assertions are a special tool to be used in rare situations where several properties or states of different components are supposed to have specific values. If we decide to put assertions in all methods like this one we'll end up with code where ```assert``` is the most used keyword in every class.

What do you think?

Reply:
My opinion:
1. `logAlreadyClosedAndReturnNull` reads "log already closed return `null`", "log already closed" sounds to me like logging is closed, "return `null`" sounds clear.
2. Yes, the method is private, but suddenly another developer does not use it correctly, otherwise he will be able to see in the log that he made a mistake and he will have to debug less.
3. For all methods it can be redundant but not terrible.

Reply:
I renamed the method to make its name less ambiguous

Reply:
How often can it be that we should return `null`?